### PR TITLE
feat: runtime algo caching, shared workspace, GQA refactor

### PR DESCRIPTION
**Pull request by zpye**
## Summary

Runtime-stage performance optimizations for hipBLASLt (MatMul, GQA) operators. Algorithm heuristics are queried once on the first inference call and cached; a lazily-grown shared GPU workspace eliminates per-call `hipMalloc`/`hipFree` overhead on subsequent runs.

### 1. Shared Workspace Infrastructure (`RuntimeState`)

- **`lib/Runtime/hipdnn_ep_runtime_state.cpp`** / **`lib/Runtime/hipdnn_ep_runtime.h`**
  - Added `workspace` (GPU pointer) and `workspace_size` fields to `RuntimeState`
  - New API: `hipdnn_ep_state_ensure_workspace(state, needed_size)` -- grows workspace if needed (never shrinks)
  - New API: `hipdnn_ep_state_get_workspace()` / `hipdnn_ep_state_get_workspace_size()` -- accessors
  - Workspace freed in `hipdnn_ep_state_cleanup()`
  - Shared across all operators (MatMul, GQA GEMMs)

### 2. MatMul Algorithm Caching (`real/matmul.cpp`)

- File-static `std::unordered_map` keyed by `(M, N, K, batch_count, elem_size)`
- On first call: `hipblasLtMatmulAlgoGetHeuristic` with 256 MB max workspace preference
- Caches optimal `algo` + `workspaceSize`; calls `ensure_workspace` to grow shared buffer
- Subsequent calls: cache hit, skip heuristic, reuse algo + shared workspace

### 3. GQA Refactor (device kernels + runtime orchestration)

- **`3rd-party/custom_kernels/hip/gqa_kernel.hip`** -- device kernels with full algorithm documentation and ONNX spec references:
  - `kv_cache_append_kernel` -- append new tokens into in-place/aliased KV cache (BSHD->BNSD transpose + scatter)
  - `kv_cache_concat_kernel` -- concatenate past data + new tokens into a fresh present buffer when past and present are separate allocations with different strides (replaces host-side B*G `hipMemcpyAsync` loop with a single kernel launch)
  - `rope_kernel` -- half-rotated RoPE
  - `transpose_mid_dims_kernel` -- 4D middle-dim transpose
  - `expand_kv_kernel` -- KV group expansion (G groups to H heads)
  - `causal_mask_kernel` -- causal attention mask
  - `softmax_inplace_kernel` -- column-wise softmax

- **`3rd-party/custom_kernels/include/hip_custom_kernels.h`** -- individual kernel launcher declarations with `present_seq` / `past_seq` parameters clarifying buffer strides

- **`lib/Runtime/real/gqa.cpp`** -- full 11-step hipBLASLt GQA pipeline:
  - hipBLASLt GEMM setup (score + value), algorithm caching per shape
  - All temp buffers packed into the shared workspace at fixed offsets, eliminating all `hipMalloc`/`hipFree` on the hot path
  - KV cache update: `kv_cache_append` for same-buffer (aliased), `kv_cache_concat` for separate-buffer with stride mismatch
  - `present_seq` used throughout (not `max_seq`) to clarify it is the present buffer's actual sequence dimension
  - `HIP_CHECK` on all kernel launcher return values with `goto cleanup` for safe resource deallocation
  - FP16-only restriction documented with future FP32/BF16 path

- **`lib/Runtime/real/error_check_macros.h`** -- `HIP_CHECK_GOTO` uses `static_cast<hipError_t>` so it works for both `hipError_t`-returning HIP APIs and `int`-returning kernel launchers

### 4. Test CMake Cleanup (`test/CMakeLists.txt`)

- Removed stale test target references

## Test Plan

- [x] Build succeeds (`_build.bat`)
- [x] Pre-commit hooks pass (trailing whitespace, clang-format, license headers)
- [x] GQA-only perf tests pass (all sequence lengths: seq1 through seq3072)
- [x] Full model perf tests pass (seq1, seq128, seq1024)

### CODE REVIEW

File: 3rd-party/custom_kernels/hip/gqa_kernel.hip [hunk 1]
```diff
@@ -7,489 +7,272 @@
...
+#include <algorithm>
 
 #include "hip_custom_kernels.h"
 
+static constexpr int GQA_THREADS = 256;
+
 // =============================================================================
 // KV Cache Update Kernel
 // =============================================================================
-// Copies current key/value into the present_key/present_value buffer at the
```
Comment (fhanuman) [inline 3rd-party/custom_kernels/hip/gqa_kernel.hip:17]: The GQA kernel I shared had detailed documentation -- ONNX spec references, BNSD layout diagrams, thread mapping, the 11-step algorithm overview. We probably missed all of it when porting it here. Could we restore all the code and algorithm documentation? The layout docs and operator spec references are especially useful for anyone debugging these later.

File: lib/Runtime/real/gqa.cpp [hunk 2]
```diff
@@ -6,19 +6,368 @@
...
 #include "hip_custom_kernels.h"
+#include "runtime_types.h"
 
+#include <algorithm>
 #include <cstdio>
+#include <cstring>
+#include <functional>
+#include <unordered_map>
+
+#define GQA_HIP_CHECK(cmd)                                                     \
```
Comment (fhanuman) [inline lib/Runtime/real/gqa.cpp]: Same issue we fixed in PR #55 -- GQA_HIP_CHECK and GQA_HIPBLAS_CHECK use return -1, which skips cleanup. In gqa_forward_hipblaslt, we create 10 hipBLASLt descriptors and have a proper cleanup: block, but any macro failure in the matmul or memcpy calls will return -1 and leak all of them. Same thing in queryGemmAlgo -- pref leaks if the set-attribute or heuristic calls fail.

The goto cleanup after the algo queries handles this correctly, but the rest of the function doesn't. Could we make all error paths go through cleanup?

File: lib/Runtime/real/gqa.cpp [hunk 3]
```diff
@@ -6,19 +6,368 @@
...
+      hip_gqa_kv_cache_update(stream, kSrc, present_key,
+                              (int)B, (int)sq, (int)G, (int)d,
+                              (int)max_seq, (int)past_len);
+      hip_gqa_kv_cache_update(stream, value, present_value,
+                              (int)B, (int)sq, (int)G, (int)d,
+                              (int)max_seq, (int)past_len);
+
+      if (past_key && past_len > 0) {
+        size_t past_bytes = (size_t)B * G * past_len * d * elem_sz;
+        GQA_HIP_CHECK(hipMemcpyAsync(present_key, past_key, past_bytes,
```
Comment (fhanuman) [inline lib/Runtime/real/gqa.cpp]: The kv_cache_update_kernel from the GQA kernel I shared assumes a single pre-allocated KV cache buffer. I might be wrong, but I think past_key and present_key can be separate buffers in ORT with different per-head strides (past_len * d vs max_seq * d). If that's the case, wouldn't the flat hipMemcpyAsync of B * G * past_len * d bytes corrupt data across head boundaries when the strides don't match?

Could we either restore the strided copy from the previous implementation or add a past_key == present_key guard here?

File: lib/Runtime/real/gqa.cpp [hunk 4]
```diff
@@ -35,10 +384,30 @@ int wrap_group_query_attention(RuntimeState *state, void *query, void *key,
...
+    fprintf(stderr, "wrap_group_query_attention: null required argument\n");
+    return -1;
+  }
+  if (num_heads % kv_num_heads != 0) {
+    fprintf(stderr, "wrap_group_query_attention: num_heads (%lld) must be "
+            "divisible by kv_num_heads (%lld)\n",
+            (long long)num_heads, (long long)kv_num_heads);
+    return -1;
+  }
+  if (element_size_bytes != 2) {
```
Comment (fhanuman) [inline lib/Runtime/real/gqa.cpp:428]: The GQA kernel I shared was FP16-only since that's what the Llama benchmarks needed, but I think the previous implementation here was templated for both float and __half. Since this rewrite now rejects FP32 entirely, is that intentional? If so, could we add a comment here noting that? Without context it could look like an oversight.

**Review by fhanuman — CHANGES_REQUESTED** [review]
The description mentions MIOpen Conv Algorithm Caching and MIOpen Tensor Descriptor Fix, but I don't see these in the final diff. It looks like the layernorm changes might have been reverted by the second commit, and the miopen caching may not have been added yet? Could we update the description to match what's actually here?

**Comment by zpye** [comment]
> The description mentions MIOpen Conv Algorithm Caching and MIOpen Tensor Descriptor Fix, but I don't see these in the final diff. It looks like the layernorm changes might have been reverted by the second commit, and the miopen caching may not have been added yet? Could we update the description to match what's actually here?

PR message has updated to describe the actual changes.

File: 3rd-party/custom_kernels/hip/gqa_kernel.hip (hunk 1 above)
Comment (zpye) [reply 3rd-party/custom_kernels/hip/gqa_kernel.hip:17]: The missing documents have been restored.

File: lib/Runtime/real/gqa.cpp (hunk 2 above)
Comment (zpye) [reply lib/Runtime/real/gqa.cpp]: Fixed. The GQA_HIP_CHECK and GQA_HIPBLAS_CHECK macros have been removed entirely. gqa.cpp now uses the shared error_check_macros.h definitions:
```
#define HIP_CHECK(cmd)     HIP_CHECK_GOTO(cmd, cleanup)
#define HIPBLAS_CHECK(cmd) HIPBLAS_CHECK_GOTO(cmd, cleanup)
```
All error paths -- including hipBLASLt calls, hipMemcpyAsync, and every kernel launcher -- now route through goto cleanup

File: lib/Runtime/real/gqa.cpp (hunk 3 above)
Comment (zpye) [reply lib/Runtime/real/gqa.cpp]: I have used two GPU kernels to solve the issue:

- kv_cache_append_kernel (same-buffer / aliased case): Only appends new tokens at [past_len, past_len+sq) via BSHD->BNSD transpose. Past data is already in place.

- kv_cache_concat_kernel (separate-buffer case): A single kernel that fills the entire present buffer -- copying past data at [0, past_len) with correct stride handling AND transposing new tokens at [past_len, past_len+sq).

The dispatch uses past_key != present_key to choose between the two paths. The concat kernel accepts separate past_seq and present_seq stride parameters to handle the mismatch correctly.

File: lib/Runtime/real/gqa.cpp (hunk 4 above)
Comment (zpye) [reply lib/Runtime/real/gqa.cpp:428]: I have added a comment here.
The main reason is all matrix layouts of hipBLASLt are created with hard-coded HIP_R_16F.
Previous implementation could be templated because the whole GQA used custom kernels. Current one should stay consistent with hipBLASLt.
FP32 support could be a future work.

**Review by fhanuman — APPROVED** [review]
